src/hashmap.c

```diff
@@ -1,51 +1,36 @@
-#define BUCKETS 1000000
-#define PRIME_SEED_1 17
-#define PRIME_SEED_2 97
-#define KEY_CHARS 128
-#define VALUE_CHARS 128
-#define NODE_SIZE sizeof(Node)
-
+#include "hashmap.h"
 #include <stdlib.h>
-#include <stdio.h>
 #include <stdbool.h>
 #include <string.h>
-#include <assert.h>
-
-/* A key-value node in a single-linked list. */
-typedef struct Node Node;
-
-struct Node {
-    Node* next;
-    char key[KEY_CHARS];
-    char value[VALUE_CHARS]; 
-};
 
+#define BUCKETS 1000000
+#define PRIME_SEED_1 17
+#define PRIME_SEED_2 97
 
+/* A key-value node in a single-linked list. */
 Node* create_sentinel_node(){
-    void* ptr = malloc(NODE_SIZE);
-    if (ptr == NULL){
+    Node* node = malloc(sizeof(Node));
+    if (!node){
         exit(EXIT_FAILURE);
-    } else {
-        Node* node = (Node*)ptr; // safely typecast the pointer
-        node->next = NULL;
-        strcpy(node->key, "\0");
-        strcpy(node->value, "\0");
-        return node;
     }
+    node->next = NULL;
+    strcpy(node->key, "\0");
+    strcpy(node->value, "\0");
+    return node;
 }
 
 
 size_t modular_hash(char* string) {
     size_t hash = 0;
     for (char* character=string; *character != '\0'; character++) {
-        hash += ((int)*character + PRIME_SEED_1) * PRIME_SEED_2; 
+        hash += (*character + PRIME_SEED_1) * PRIME_SEED_2; 
     }
     return hash;
 }
 
 
 Node** hm_create(void){
-    static Node* keys[BUCKETS] = {};
+    static Node* keys[BUCKETS] = {}; // this makes the hashmap a singleton. Use malloc instead
```

Comment:
Having mutable static variables inside of a function is almost always awful. Here the table is allocated on program startup, not when `hm_create` is called, and so no matter how many times you call `hm_create` it will always return the same hashmap. What you probably want is for `hm_create` to create a new table which has been allocated with malloc each time it is called

Reply:
Yeah, I realised this too - malloc is used now.